src/scores/categorical/contingency_impl.py

```diff
@@ -0,0 +1,263 @@
+    weights: xr.DataArray = None,
+):
+    """
+    Functional API example for contingency scores
+
+    If the event operator is None, then fcst and obs are held to contain
+    binary information about whether an event is predicted or occurred.
+    If the event operator is provided, it will be utilised to produce the
+    event tables for the forecast and observed conditions prior to generating
+    the contingency table.
+    """
+
+    if event_operator is not None:
+        table = event_operator.make_table(fcst, obs)
+    else:
+        table = BinaryContingencyTable(fcst, obs)
+
+    acc = table.accuracy(reduce_dims=reduce_dims, preserve_dims=preserve_dims)
+    return acc
+
+
+class BinaryContingencyTable:
+    """
+    At each location, the value will either be:
+     - A true positive    (0)
+     - A false positive   (1)
+     - A true negative    (2)
+     - A false negative   (3)
+
+    It will be common to want to operate on masks of these values,
+    such as:
+     - Plotting these attributes on a map
+     - Calculating the total number of these attributes
+     - Calculating various ratios of these attributes, potentially
+       masked by geographical area (e.g. accuracy in a region)
+
+    As such, the per-pixel information is useful as well as the overall
+    ratios involved.
+    """
+
+    def __init__(self, forecast_events, observed_events):
+        # type checking goes here
+        self.forecast_events = forecast_events
+        self.observed_events = observed_events
+
+        self.tp = (self.forecast_events == 1) & (self.observed_events == 1)  # true positives
+        self.tn = (self.forecast_events == 0) & (self.observed_events == 0)  # true negatives
+        self.fp = (self.forecast_events == 1) & (self.observed_events == 0)  # false positives
+        self.fn = (self.forecast_events == 0) & (self.observed_events == 1)  # false negatives
+
+        # Variables for count-based metrics, calculated on first access
+        self.tp_count = self.tp.sum()  # Count of true positives
+        self.tn_count = self.tn.sum()  # Count of true negatives
+        self.fp_count = self.fp.sum()  # Count of false positives
+        self.fn_count = self.fn.sum()  # Count of true negatives
+        self.total_count = self.tp_count + self.tn_count + self.fp_count + self.fn_count
+
+    def generate_counts(self, *, reduce_dims=None, preserve_dims=None):
+        to_reduce = scores.utils.gather_dimensions(
+            self.forecast_events.dims, self.observed_events.dims, reduce_dims=reduce_dims, preserve_dims=preserve_dims
+        )
+
+        cd = {
+            "tp_count": self.tp.sum(to_reduce),
+            "tn_count": self.tn.sum(to_reduce),
+            "fp_count": self.fp.sum(to_reduce),
+            "fn_count": self.fn.sum(to_reduce),
+        }
+        total = cd["tp_count"] + cd["tn_count"] + cd["fp_count"] + cd["fn_count"]
+        cd["total_count"] = total
+
+        return cd
```

Comment:
[optional] 
Could you explain the decision to return this rather than store it as a class member?

I guess if you want to re-use the object to repeat the computations for different `preserve_dims` & `reduce_dims` it does make sense; still might be worth caching `generate_counts` e.g. using `functools.cache`, since it's getting re-used a lot.

Reply:
It is now cached. Thanks.